mettle/worker/redis.go

```diff
@@ -91,23 +94,33 @@ func (r *redisClient) Healthcheck() error {
 func (r *redisClient) ReadBlob(dg *pb.Digest) ([]byte, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), r.timeout)
 	defer cancel()
-	cmd := r.readRedis.Get(ctx, dg.Hash)
-	if err := cmd.Err(); err == nil {
-		blob, _ := cmd.Bytes()
-		return blob, nil
-	} else if err != redis.Nil {
-		log.Warning("Failed to read blob from Redis: %s", err)
+
+	if r.limiter.Allow() {
+		cmd := r.readRedis.Get(ctx, dg.Hash)
+		if err := cmd.Err(); err == nil {
+			blob, _ := cmd.Bytes()
+			return blob, nil
+		} else if err != redis.Nil {
+			log.Warningf("Failed to read blob from Redis: %s", err)
+			r.limiter.Reserve()
+		}
+	} else {
+		log.Warningf("limiter error rate exceeded, skipping Redis lookup")
 	}
 	// If we get here, the blob didn't exist. Download it then write back to Redis.
 	blob, err := r.elan.ReadBlob(dg)
 	if err != nil {
 		return nil, err
 	}
 	if dg.SizeBytes < r.maxSize {
-		ctx, cancel = context.WithTimeout(context.Background(), r.timeout)
-		defer cancel()
-		if cmd := r.redis.Set(ctx, dg.Hash, blob, 0); cmd.Val() != "OK" {
-			log.Warning("Failed to set blob in Redis: %s", cmd.Err())
+		if r.limiter.Allow() {
+			go func() {
+				ctx, cancel = context.WithTimeout(context.Background(), r.timeout)
+				defer cancel()
+				if cmd := r.redis.Set(ctx, dg.Hash, blob, 0); cmd.Val() != "OK" {
```

Comment:
One thing to be mildly cautious of here is that `dg` is now used after this function returns. I doubt this is likely to be a problem anywhere, but just mentioning.

Reply:
I'll close over the `dg.Hash` just to remove any doubt!